internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+	"github.com/tikv/client-go/v2/util"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
```

Comment:
Is there a specific reason to pass a flush func from outside rather than making it a method or function defined inside this module? From my perspective, it is more obscure to understand which variables are read by the flush func, and which variables it modifies. It makes understanding the concurrency correctness a bit harder (at least for me)?

Reply:
There are 2 reasons of passing a flush func:

1. I'm hesitate to introduce RPC calling code into the unionstore module, it sounds like an in-mem module.
2. If there are some other implementation of flush func(like S3 based flush), we can quickly replace it, so the flush func is a plugin module for pipelined memdb instead of core logic.